### PR TITLE
ibis implementation

### PR DESCRIPTION
**Pull request by ian-r-rose**
Partner to intake/intake#395

Given a SQL data source and a table, we can produce ibis clients for constructing lazily evaluated expressions. Currently targeting the ibis backends that use SQLAlchemy, so PostgreSQL, SQLite, and MySQL (I've tested the former two).

### CODE REVIEW

File: intake_sql/sql_cat.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def _load(self):
                 if c.primary_key:
                     description = 'SQL table %s from %s' % (name, self.uri)
                     args = {'uri': self.uri, 'table': name, 'index': c.name,
-                            'sql_kwargs': self.kwargs}
+                            'sql_kwargs': {}}
```
Comment (martindurant) [inline intake_sql/sql_cat.py]: Why this change, should we not be able to pass kwargs to the loader?

File: intake_sql/intake_sql.py [hunk 2]
```diff
@@ -224,6 +249,18 @@ def to_dask(self):
...
 
+    def to_ibis(self):
+        """
+        Create an ibis expression for the data source.
+        The sql_expr for the source must be a table, not a table expression.
+        The ibis expression is not partitioned.
+        """
+        client = make_ibis_client(self._uri)
+        if self._sql_expr not in client.list_tables():
+            raise ValueError("Only full tables can be used in to_ibis")
```
Comment (martindurant) [inline intake_sql/intake_sql.py]: I suppose the table expression might be indeed a whole table (or view) - can that be found out, or do you wish to strictly keep to string table names?

File: intake_sql/sql_cat.py (hunk 1 above)
Comment (ian-r-rose) [reply intake_sql/sql_cat.py]: I ran into troubles here testing with SQLite.  `self.kwargs` included both `"name"` and `"ttl"`, neither of which were valid kwargs for SQLAlchemy. It's not clear to me what the intent of these `kwargs` was, but that seemed overly broad to me.

Perhaps driver should take a `sql_kwargs` argument in its `args`, which will be passed to the table reader?

File: intake_sql/intake_sql.py (hunk 2 above)
Comment (ian-r-rose) [reply intake_sql/intake_sql.py]: This is a limitation of the ibis SQLAlchemy-based clients. Some backends have been set up to use an expression as a subquery, but not these.

Here is an example of a client without such a restriction: https://github.com/Quansight/intake-omnisci/blob/089beb0de75d021221791cc49796ed8e8ac20456/intake_omnisci/source.py#L124-L127

File: intake_sql/sql_cat.py (hunk 1 above)
Comment (martindurant) [reply intake_sql/sql_cat.py]: Ah, I see that this was fixed in my local branch but never pushed. What do you think of this:
```diff
--- a/intake_sql/sql_cat.py
+++ b/intake_sql/sql_cat.py
@@ -15,7 +15,8 @@ class SQLCatalog(Catalog):
     name = 'sql_cat'
     version = __version__

-    def __init__(self, uri, views=False, **kwargs):
+    def __init__(self, uri, views=False, sql_kwargs=None, **kwargs):
+        self.sql_kwargs = sql_kwargs or {}
         self.uri = uri
         self.views = views
         super(SQLCatalog, self).__init__(**kwargs)
@@ -32,7 +33,7 @@ class SQLCatalog(Catalog):
                 if c.primary_key:
                     description = 'SQL table %s from %s' % (name, self.uri)
                     args = {'uri': self.uri, 'table': name, 'index': c.name,
-                            'sql_kwargs': self.kwargs}
+                            'sql_kwargs': self.sql_kwargs}
                     e = LocalCatalogEntry(name, description, 'sql_auto', True,
                                           args, {}, {}, {}, "", getenv=False,
                                           getshell=False)
```

We generally want to allow general kwargs which may be ignored or picked up by the base class, to allow for sources accessed via a server, for example.

**Comment by ian-r-rose** [comment]
Agreed that tests would be useful here. We can test against SQLite without too much difficulty, I think. In principle, Postgres and MySQL can also be tested using their docker containers, but that may be more effort than it's worth.

File: intake_sql/sql_cat.py (hunk 1 above)
Comment (ian-r-rose) [reply intake_sql/sql_cat.py]: Yes, this diff looks right to me. If you push it I can rebase this PR against it.

Comment (ian-r-rose) [reply intake_sql/sql_cat.py]: Added this diff, but I can deal with the conflict if you prefer to push your commit.

**Comment by ian-r-rose** [comment]
Added a (very) simple test for this behavior.

**Comment by martindurant** [comment]
OK, I think this is a good start, and we'll see how people use it.
